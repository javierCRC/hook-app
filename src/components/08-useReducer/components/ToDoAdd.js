import React from 'react'
import { useForm } from '../../../hooks/useForm';
import { f_MessageProcess } from '../../../helpers';

export const ToDoAdd = ({ pF_handleAddTodo }) => {

    // custom hook for capture the value of each input of any form.
    const { stateValues , f_handleInputChange , f_handleCleanForm } = useForm({
        txtDescription:''
    });

    // destructuring the description value of stateValue for be use to individual form
    const { txtDescription } = stateValues;
    
    //console.log(txtDescription);

    const f_handleSubmitAddTodo = (e) => {
        e.preventDefault();

        if(txtDescription.trim().length <= 1){
           f_MessageProcess("Error empty field","Can't be empty fields","error");
           return;
        }

        const c_NewTodo = {
           id : new Date().getTime(),
           desc: txtDescription,
           done: false
       };

       
       pF_handleAddTodo ( c_NewTodo ); // sending the action to the reducer 

       f_handleCleanForm(); // for clean the form values before use the custom hook "useForm"

   }

    return (
        <>
                 <h4> Add ToDo <i className="far fa-list-alt"></i> </h4>
                                <hr />

                                <form    onSubmit={ f_handleSubmitAddTodo }
                                        className="bg-light my-3 p-3 pt-lg-0 border rounded">
                                   <h5 className="font-weight-bold text-center my-4 text-success"> Registry <i className="fas fa-calendar-day"></i></h5>     
                                
                                    <div className="form-group form-row">
                                        <label  className="col-12 col-form-label">Description:</label>

                                        <div className="input-group col-12 col-md-10">
                                            <div className="input-group-prepend">
                                                <span className="input-group-text text-primary"><i className="fas fa-id-card"></i></span>
                                            </div>
                                            <input type="text" 
                                                   name='txtDescription'
                                                   className="form-control"
                                                   autoComplete='off' 
                                                   placeholder="Whrite a new To Do" 
                                                   value={ txtDescription }
                                                   onChange={ f_handleInputChange }
                                                   />
                                        </div>
                                    </div>

                                    <div className="form-group form-row mb-0">
                                        <div className="col-auto">
                                            <button type="submit" 
                                                    className="btn btn-primary btn-block"
                                            > Save <i className="fas fa-save"></i>
                                            </button>
                                        </div>
                                    </div>

                                </form>
        
            
        </>
    )
}
