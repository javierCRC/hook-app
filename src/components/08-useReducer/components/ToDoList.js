import React from 'react'
import { ToDoListItem } from '../../08-useReducer/components/ToDoListItem';

export const ToDoList = ( { pTodo = [] , pF_handleToggleTask, pF_f_handleClickDelete } ) => {
    return (
        <ul>
            {
                                        
                   pTodo.map( ( vTodo , i) => ( 
                    <ToDoListItem 
                                   key = { pTodo.id } 
                                   pToDo2={ vTodo } 
                                   pIndex={ i } 
                                   pF_handleToggleTask2={ pF_handleToggleTask } 
                                   pF_f_handleClickDelete2={ pF_f_handleClickDelete }
                    />
                ))
            }   
        </ul>
    ) 
}
