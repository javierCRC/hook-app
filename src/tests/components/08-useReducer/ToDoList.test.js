import { shallow } from "enzyme";

import { arrToDos } from "../../fixtures/demoToDos";
import { ToDoList } from '../../../components/08-useReducer/components/ToDoList';
import '@testing-library/jest-dom'; 


describe('Pruebas de <ToDoList >', () => {
    
     
    const vF_handleToggleTask = jest.fn(); 
    const vf_handleClickDelete = jest.fn();
 
     const cWrapper = shallow(<ToDoList pTodo = { arrToDos  } 
                                                    pF_handleToggleTask = { vF_handleToggleTask }
                                                    pF_f_handleClickDelete = { vf_handleClickDelete }
                                                    />
                                     );
    
    test('1.Must do render the component ToDoList.', () => {
        
        // 1 Arrange *******************************************************************************
            
        //2. Act *******************************************************************************
            //console.log(cWrapper.html());

        //3. Assert *******************************************************************************
             expect(cWrapper).toMatchSnapshot();

    });

    test('2.Must be have two <ToDoListItem />.', () => {
        
        // 1 Arrange *******************************************************************************
             let vCountToDoListItem = 0;
             let vCountToDoListItemExpected = arrToDos.length;

        //2. Act *******************************************************************************
             vCountToDoListItem = cWrapper.find('ToDoListItem').length;
               //console.log('count: '+ vCountToDoListItem );  

        //3. Assert *******************************************************************************
             expect(vCountToDoListItem).toBe(vCountToDoListItemExpected);
            
    });

    test('2.Must be have the function ToDoListItem -> pF_f_handleClickDelete2.', () => {
        
        // 1 Arrange *******************************************************************************
             let vHaveDeleteFunction;
             let vFunctionExpected = expect.any(Function);

        //2. Act *******************************************************************************
              vHaveDeleteFunction = cWrapper.find('ToDoListItem').at(0).prop('pF_f_handleClickDelete2');

               //console.log('Delete Function: '+ vHaveDeleteFunction );  

        //3. Assert *******************************************************************************
             expect(vHaveDeleteFunction).toEqual( vFunctionExpected );
            
    });

    test('3.Must be have the function ToDoListItem -> pF_handleToggleTask2.', () => {
        
        // 1 Arrange *******************************************************************************
             let vHaveToggleFunction;
             let vFunctionExpected = expect.any(Function);
             
        //2. Act *******************************************************************************
              vHaveToggleFunction = cWrapper.find('ToDoListItem').at(0).prop('pF_handleToggleTask2');

               //console.log('Toggle Function: '+ vHaveToggleFunction );  

        //3. Assert *******************************************************************************
             expect(vHaveToggleFunction).toEqual( vFunctionExpected );
            
    });

    
})
