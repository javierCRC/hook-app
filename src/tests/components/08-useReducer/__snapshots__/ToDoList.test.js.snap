// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <ToDoList > 1.Must do render the component ToDoList. 1`] = `
<ul>
  <ToDoListItem
    pF_f_handleClickDelete2={[MockFunction]}
    pF_handleToggleTask2={[MockFunction]}
    pIndex={0}
    pToDo2={
      Object {
        "desc": "Learn React with Js",
        "done": false,
        "id": 1,
      }
    }
  />
  <ToDoListItem
    pF_f_handleClickDelete2={[MockFunction]}
    pF_handleToggleTask2={[MockFunction]}
    pIndex={1}
    pToDo2={
      Object {
        "desc": "Learn React with Ts",
        "done": true,
        "id": 2,
      }
    }
  />
</ul>
`;
